File: chrome/browser/download/android/rename_utils.cc
// Copyright 2019 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "base/android/jni_string.h"
#include "base/files/file_path.h"
#include "chrome/android/chrome_jni_headers/RenameUtils_jni.h"

using base::android::ConvertUTF8ToJavaString;
using base::android::JavaParamRef;
using base::android::ScopedJavaLocalRef;

// static
static base::android::ScopedJavaLocalRef<jstring>
JNI_RenameUtils_GetFileExtension(
    JNIEnv* env,
    const base::android::JavaParamRef<jstring>& fileName) {
  std::string extension =
      base::FilePath(ConvertJavaStringToUTF8(env, fileName)).Extension();
  return ConvertUTF8ToJavaString(env, extension);
}

File: chrome/android/java/src/org/chromium/chrome/browser/download/home/rename/RenameUtils.java
// Copyright 2019 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.
package org.chromium.chrome.browser.download.home.rename;

import androidx.annotation.VisibleForTesting;

import org.chromium.base.FileUtils;
import org.chromium.base.annotations.NativeMethods;

/**
 * A class containing some utility static methods for rename.
 */
public class RenameUtils {
    private static boolean sIsDisabledNativeForTesting;

    /**
     * Determine the extension of a downloaded item.
     *
     * @param fileName input to determine the extension.
     * @return String representing the extension part of fileName.
     * Return empty if there is no extension, otherwise return extension part starting with
     * separator. e.g. it will return ".jpg" for the path "foo/bar.jpg", and return ".tar.gz" for
     * the path "foo/bar.tar.gz".
     */
    public static String getFileExtension(String fileName) {
        return sIsDisabledNativeForTesting ? FileUtils.getExtension(fileName)
                                           : RenameUtilsJni.get().getFileExtension(fileName);
    }

    /**
     * Disables the native APIs. This is only intended for testing purposes.
     */
    @VisibleForTesting
    public static void disableNativeForTesting() {
        sIsDisabledNativeForTesting = true;
    }

    @NativeMethods
    interface Natives {
        String getFileExtension(String fileName);
    }
}
